urlce7759b8da_Palachsuper_EBA4_tgz-pJ8-json_TestJ8/src/terminal/devices/Device.java

```diff
@@ -0,0 +1,376 @@
+                    cashe_machine.GenerateJSONAnswer("500");
+                    break;
+                case p0203078106624f0000000000:
+                    isBill=true;
+                    cashe_machine.isBillGetMoney=false;
+                    Service.WritePaymentLog("Вставлена 1000 рублей",cashe_machine.id_payment);
+                    cashe_machine.GenerateJSONAnswer("1000");
+                    break;
+                case p0203078107eb5e0000000000:
+                    isBill=true;
+                    cashe_machine.isBillGetMoney=false;
+                    Service.WritePaymentLog("Вставлены 5000 рублей",cashe_machine.id_payment);
+                    cashe_machine.GenerateJSONAnswer("5000");
+                    break;
+                case p0203071c6077e00000000000:
+                    //Service.WritePaymentLog("прочитали плохую купюру",cashe_machine.id_payment);
+                    cashe_machine.Bad_cashe();
+                    cashe_machine.col_bad=200;
+                    break;
+                case p0203061467d4000000000000:
+                    cashe_machine.col_bad=0;
+                    break;
+            }
+        } catch (IllegalArgumentException iae) {
+            /// опа, мы получили неизвестное значение. Можно присвоить
+            // некое default и продолжить обрабатывать его в case, а можно бросить
+            // наверх
+        }
+    }
+
+	// запуск потока устройства
+	public void run()
+	{
+		try {
+			this.connect();
+		} catch (Exception e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+	}
+	
+	// инициализация устройства
+	public void init(JSONObject data)
+	{
+		String msg = "";
+		
+		try 
+		{
+			if (this.connect())
+			{
+				msg += "Connected (port="+this.port+"; rate="+this.portRate+"; parity="+this.parity+"; rts="+this.rts+"). ";
+				msg += this.initDetails(data);
+			}
+			else
+			{
+				this.error = "Not connected (port="+this.port+"; rate="+this.portRate+"; parity="+this.parity+"; rts="+this.rts+")";
+			}
+		} 
+		catch (Exception e)
+		{
+			this.error = "Not connected (port="+this.port+"; rate="+this.portRate+"; parity="+this.parity+"; rts="+this.rts+")";
+		}
+		
+		this.method = "onInit"; // callback в javascript->DeviceConnector
+	}
+	
+	// дополнительная реализация инициализации устройства (см. в потомках)
+	public String initDetails(JSONObject data)
+	{
+		return "";
+	}
+	
+	public void setPort(String port)
+	{
+		this.port = port;
+	}
+	
+	public void setPortRate(int portRate)
+	{
+		this.portRate = portRate;
+	}
+
+    public static String getPropPort()
+    {
+        String dat_time_path="";
+        StringBuilder sb = new StringBuilder();
+        try
+        {
+
+            try
+            {    dat_time_path= "/home/terminal/lib/prop_port.txt";
+                BufferedReader in = new BufferedReader(new FileReader( new File(dat_time_path).getAbsoluteFile()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *in*. The resource is referenced at lines 196, 198. The resource is closed at line 202. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).